File: content/develop/use-cases/time-series-dashboard/redis-py/sensor_simulator.py
"""
Simulated sensor definitions and value generation for the dashboard demo.
"""

from __future__ import annotations

from dataclasses import dataclass
import random


@dataclass(frozen=True)
class SensorDefinition:
    """Describe one simulated power sensor and its Redis labels."""

    sensor_id: str
    zone: str
    unit: str
    min_value: float
    max_value: float
    baseline: float
    drift: float
    spike_chance: float = 0.0
    spike_size: float = 0.0

    @property
    def sensor_type(self) -> str:
        return "power_consumption"

    @property
    def key(self) -> str:
        return f"ts:sensor:{self.sensor_type}:{self.sensor_id}"

    @property
    def labels(self) -> dict[str, str]:
        return {
            "site": "demo",
            "sensor_type": self.sensor_type,
            "sensor_id": self.sensor_id,
            "zone": self.zone,
            "unit": self.unit,
        }


SENSORS: list[SensorDefinition] = [
    SensorDefinition(
        sensor_id="power-1",
        zone="north",
        unit="watts",
        min_value=320.0,
        max_value=920.0,
        baseline=480.0,
        drift=18.0,
        spike_chance=0.10,
        spike_size=120.0,
    ),
    SensorDefinition(
        sensor_id="power-2",
        zone="central",
        unit="watts",
        min_value=320.0,
        max_value=920.0,
        baseline=560.0,
        drift=20.0,
        spike_chance=0.08,
        spike_size=140.0,
    ),
    SensorDefinition(
        sensor_id="power-3",
        zone="south",
        unit="watts",
        min_value=320.0,
        max_value=920.0,
        baseline=640.0,
        drift=24.0,
        spike_chance=0.12,
        spike_size=160.0,
    ),
]


class SensorSimulator:
    """Generate smooth, bounded readings for a set of sensors."""

    def __init__(self, sensors: list[SensorDefinition]) -> None:
        self.sensors = sensors
        self._values = {sensor.sensor_id: sensor.baseline for sensor in sensors}

    def next_samples(self) -> list[tuple[SensorDefinition, float]]:
        samples: list[tuple[SensorDefinition, float]] = []

        for sensor in self.sensors:
            current = self._values[sensor.sensor_id]
            drift = random.uniform(-sensor.drift, sensor.drift)
            pull = (sensor.baseline - current) * 0.12
            value = current + drift + pull

            if sensor.spike_chance and random.random() < sensor.spike_chance:
                value += random.uniform(0.5, 1.0) * sensor.spike_size

            value = max(sensor.min_value, min(sensor.max_value, value))
            value = round(value, 2)

            self._values[sensor.sensor_id] = value
            samples.append((sensor, value))

        return samples

File: content/develop/use-cases/time-series-dashboard/java-jedis/SensorSimulator.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Simulated sensor definitions and value generation for the dashboard demo.
 */
public final class SensorSimulator {
    public record SensorDefinition(
            String sensorId,
            String zone,
            String unit,
            double minValue,
            double maxValue,
            double baseline,
            double drift,
            double spikeChance,
            double spikeSize) {

        public String sensorType() {
            return "power_consumption";
        }

        public String key() {
            return "ts:sensor:" + sensorType() + ":" + sensorId;
        }

        public Map<String, String> labels() {
            Map<String, String> labels = new LinkedHashMap<>();
            labels.put("site", "demo");
            labels.put("sensor_type", sensorType());
            labels.put("sensor_id", sensorId);
            labels.put("zone", zone);
            labels.put("unit", unit);
            return labels;
        }
    }

    public record SensorSample(SensorDefinition sensor, double value) {
    }

    public static final List<SensorDefinition> SENSORS = List.of(
            new SensorDefinition("power-1", "north", "watts", 320.0, 920.0, 480.0, 18.0, 0.10, 120.0),
            new SensorDefinition("power-2", "central", "watts", 320.0, 920.0, 560.0, 20.0, 0.08, 140.0),
            new SensorDefinition("power-3", "south", "watts", 320.0, 920.0, 640.0, 24.0, 0.12, 160.0)
    );

    private final List<SensorDefinition> sensors;
    private final Map<String, Double> values = new LinkedHashMap<>();

    public SensorSimulator(List<SensorDefinition> sensors) {
        this.sensors = sensors;
        for (SensorDefinition sensor : sensors) {
            values.put(sensor.sensorId(), sensor.baseline());
        }
    }

    public List<SensorSample> nextSamples() {
        List<SensorSample> samples = new ArrayList<>();

        for (SensorDefinition sensor : sensors) {
            double current = values.get(sensor.sensorId());
            double drift = randomBetween(-sensor.drift(), sensor.drift());
            double pull = (sensor.baseline() - current) * 0.12;
            double value = current + drift + pull;

            if (sensor.spikeChance() > 0 && ThreadLocalRandom.current().nextDouble() < sensor.spikeChance()) {
                value += randomBetween(0.5, 1.0) * sensor.spikeSize();
            }

            value = Math.max(sensor.minValue(), Math.min(sensor.maxValue(), value));
            value = Math.round(value * 100.0) / 100.0;

            values.put(sensor.sensorId(), value);
            samples.add(new SensorSample(sensor, value));
        }

        return samples;
    }

    private double randomBetween(double min, double max) {
        return ThreadLocalRandom.current().nextDouble(min, max);
    }
}
